<?xml version="1.0"?>
<robot name="braccio">
     <gazebo>
     <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
       <robotNamespace>/robot</robotNamespace>
       <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
       <controlPeriod>0.001</controlPeriod>
       <legacyModeNS>true</legacyModeNS>
     </plugin>
   </gazebo>
    <!-- Color de materiales -->
    <material name="orange">
        <color rgba="0.57 0.17 0.0 1" />
    </material>
    <material name="white">
        <color rgba="0.8 0.8 0.8 1.0" />
    </material>
    <material name="black">
        <color rgba="0 0 0 0.50" />
    </material>

    <!--Mascaras para partes del robot-->
    <link name="base_link">
        <visual>
            <geometry>
                <cylinder length="0.01" radius=".053" />
            </geometry>
            <material name="black" />
            <origin rpy="0 0 0" xyz="0 0 0" />
        </visual>
        <inertial>
            <origin rpy="0 0 0" xyz="0 0 0" />
            <mass value="2" />
            <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.00666666666667" iyz="0" izz="0.0216666666667" />
        </inertial>
    </link>

    <link name="braccio_base_link">
        <visual>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_base.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <material name="orange" />
            <origin rpy="0 0 -1.5708" xyz="0 0.004 0" />
        </visual>
        <collision>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_base.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <origin rpy="0 0 -1.5708" xyz="0 0.004 0" />
        </collision>
        <inertial>
            <origin rpy="0 0 3.1416" xyz="0 0.004 0" />
            <mass value="2" />
            <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.00666666666667" iyz="0" izz="0.0216666666667" />
        </inertial>
    </link>

    <link name="shoulder_link">
        <visual>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_shoulder.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <material name="orange" />
            <origin rpy="0 0 0" xyz="-0.01 0.0056 -0.021" />
        </visual>
        <collision>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_shoulder.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <origin rpy="0 0 0" xyz="-0.01 0.0056 -0.021" />
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="-0.01 0.0056 -0.021" />
            <mass value="0.1" />
            <inertia ixx="0.000348958333333" ixy="0" ixz="0" iyy="0.000348958333333" iyz="0" izz="3.125e-05" />
        </inertial>
    </link>

    <link name="elbow_link">
        <visual>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_elbow.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <material name="orange" />
            <origin rpy="0 0 0" xyz="-0.01 0 -0.005" />
        </visual>
        <collision>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_elbow.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <origin rpy="0 0 0" xyz="-0.01 0 -0.005" />
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="-0.01 0 -0.005" />
            <mass value="0.1" />
            <inertia ixx="0.000348958333333" ixy="0" ixz="0" iyy="0.000348958333333" iyz="0" izz="3.125e-05" />
        </inertial>
    </link>

    <link name="wrist_pitch_link">
        <visual>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_wrist_pitch.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <material name="orange" />
            <origin rpy="0 0 0" xyz="0.003 -0.0015 0.0085" />
        </visual>
        <collision>
            <geometry>
                 <mesh filename="package://lab2/meshes/braccio_wrist_pitch.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <origin rpy="0 0 0" xyz="0.003 -0.0015 0.0085" />
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="0.003 -0.0015 0.0085" />
            <mass value="0.1" />
            <inertia ixx="0.000348958333333" ixy="0" ixz="0" iyy="0.000348958333333" iyz="0" izz="3.125e-05" />
        </inertial>
    </link>

    <link name="wrist_roll_link">
        <visual>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_wrist_roll.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <material name="white" />
            <origin rpy="0 0 0" xyz="0.006 0 0.035" />
        </visual>
        <collision>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_wrist_roll.stl" scale="0.01 0.01 0.01" />
            </geometry>
            <origin rpy="0 0 0" xyz="0.006 0 0.035" />
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="0.006 0 0.035" />
            <mass value="0.1" />
            <inertia ixx="0.000348958333333" ixy="0" ixz="0" iyy="0.000348958333333" iyz="0" izz="3.125e-05" />
        </inertial>
    </link>

    <link name="left_gripper_link">
        <visual>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_left_gripper.stl" scale="0.001 0.001 0.001" />
            </geometry>
            <material name="white" />
            <origin rpy="0 1.5708 0" xyz="0.035 -0.0215 0" />
        </visual>
        <collision>
            <origin rpy="0 1.5708 0" xyz="0.035 -0.0215 0" />
            <geometry>
                 <mesh filename="package://lab2/meshes/braccio_left_gripper.stl" scale="0.001 0.001 0.001" />
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1" />
            <inertia ixx="0.000348958333333" ixy="0" ixz="0" iyy="0.000348958333333" iyz="0" izz="3.125e-05" />
        </inertial>
    </link>

    <link name="right_gripper_link">
        <visual>
            <geometry>
                <mesh filename="package://lab2/meshes/braccio_right_gripper.stl" scale="0.001 0.001 0.001" />
            </geometry>
            <material name="white" />
            <origin rpy="0 1.5708 0" xyz="0.035 -0.0225 0.010" />
        </visual>
        <collision>
            <origin rpy="0 1.5708 0" xyz="0.035 -0.0225 0.010" />
            <geometry>
                 <mesh filename="package://lab2/meshes/braccio_right_gripper.stl" scale="0.001 0.001 0.001" />
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1" />
            <inertia ixx="0.000348958333333" ixy="0" ixz="0" iyy="0.000348958333333" iyz="0" izz="3.125e-05" />
        </inertial>
    </link>

    <!--Articulaciones del robot-->

    <joint name="base_joint" type="revolute">
        <axis xyz="0 0 -1" />
        <limit effort="1000.0" lower="0.0" upper="3.1416" velocity="1.0" />
        <origin rpy="0 0 0" xyz="0 0 0" />
        <parent link="base_link" />
        <child link="braccio_base_link" />
    </joint>
    <transmission name="base_trasmision">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="base_joint">
            <hardwareInterface>
                hardware_interface/PositionJointInterface
            </hardwareInterface>
        </joint>
        <actuator name="base_motor">
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <joint name="shoulder_joint" type="revolute">
        <axis xyz="1 0 0" />
        <limit effort="1000.0" lower="0.2618" upper="2.8798" velocity="1.0" />
        <origin rpy="0 0 -1.5708" xyz="0 -.002 0.072" />
        <parent link="braccio_base_link" />
        <child link="shoulder_link" />
    </joint>
    <transmission name="shoulder_trasmision">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="shoulder_joint">
            <hardwareInterface>
                hardware_interface/PositionJointInterface
            </hardwareInterface>
        </joint>
        <actuator name="shoulder_motor">
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <joint name="elbow_joint" type="revolute">
        <axis xyz="1 0 0" />
        <limit effort="1000.0" lower="0" upper="3.1416" velocity="1.0" />
        <origin rpy="0 0 0" xyz="0 0 0.125" />
        <parent link="shoulder_link" />
        <child link="elbow_link" />
    </joint>
    <transmission name="elbow_trasmision">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="elbow_joint">
            <hardwareInterface>
                hardware_interface/PositionJointInterface
            </hardwareInterface>
        </joint>
        <actuator name="elbow_motor">
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <joint name="wrist_pitch_joint" type="revolute">
        <axis xyz="1 0 0" />
        <limit effort="1000.0" lower="0" upper="3.1416" velocity="1.0" />
        <origin rpy="0 0 0" xyz="0 0 0.125" />
        <parent link="elbow_link" />
        <child link="wrist_pitch_link" />
    </joint>
    <transmission name="wrist_pitch_trasmision">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_pitch_joint">
            <hardwareInterface>
                hardware_interface/PositionJointInterface
            </hardwareInterface>
        </joint>
        <actuator name="wrist_pitch_motor">
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <joint name="wrist_roll_joint" type="revolute">
        <axis xyz="0 0 -1" />
        <limit effort="1000.0" lower="0.0" upper="3.1416" velocity="1.0" />
        <origin rpy="0 0 1.5708" xyz="0 0.0 0.06" />
        <parent link="wrist_pitch_link" />
        <child link="wrist_roll_link" />
    </joint>
    <transmission name="wrist_roll_trasmision">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_roll_joint">
            <hardwareInterface>
                hardware_interface/PositionJointInterface
            </hardwareInterface>
        </joint>
        <actuator name="wrist_roll_motor">
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <joint name="gripper_joint" type="revolute">
        <axis xyz="0 -1 0" />
        <limit effort="1000.0" lower="0.95" upper="1.54" velocity="1.0" />
        <origin rpy="0 0 0" xyz="0.010 0 0.03" />
        <parent link="wrist_roll_link" />
        <child link="right_gripper_link" />
    </joint>
    <transmission name="gripper_trasmision">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="gripper_joint">
            <hardwareInterface>
                hardware_interface/PositionJointInterface
            </hardwareInterface>
        </joint>
        <actuator name="gripper_motor">
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    

    <joint name="sub_gripper_joint" type="revolute">
        <axis xyz="0 1 0" />
        <mimic joint="gripper_joint" />
        <limit effort="1000.0" lower="0.95" upper="1.54" velocity="1.0" />
        <origin rpy="0 3.1416 0" xyz="-0.010 0 0.03" />
        <parent link="wrist_roll_link" />
        <child link="left_gripper_link" />
    </joint>
    <transmission name="sub_gripper_trasmision">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="sub_gripper_joint">
            <hardwareInterface>
                hardware_interface/PositionJointInterface
            </hardwareInterface>
        </joint>
        <actuator name="sub_gripper_motor">
        <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    
    <link name="world"/>
    <joint name="world_joint" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
    </joint>
    
    <gazebo reference="base_link">
        <mu1>1</mu1>
        <mu2>1</mu2>
        <selfCollide>true</selfCollide>
        <material>Gazebo/black</material>
    </gazebo>
    
    <gazebo reference="gripper_joint">
        <implicitSpringDamper>1</implicitSpringDamper>
    </gazebo>
    <gazebo reference="sub_gripper_joint">
     <implicitSpringDamper>1</implicitSpringDamper>
    </gazebo>

  <gazebo reference="braccio_base_link">
    <material>Gazebo/Orange</material>
    <mu1>1</mu1>
    <mu2>1</mu2>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="shoulder_link">
    <material>Gazebo/Orange</material>
    <mu1>1</mu1>
    <mu2>1</mu2>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="elbow_link">
    <material>Gazebo/Orange</material>
    <mu1>1</mu1>
    <mu2>1</mu2>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_pitch_link">
    <material>Gazebo/Orange</material>
    <mu1>1</mu1>
    <mu2>1</mu2>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_roll_link">
    <material>Gazebo/Orange</material>
    <mu1>1</mu1>
    <mu2>1</mu2>
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="right_gripper_link">
    <material>Gazebo/White</material>
    <selfCollide>true</selfCollide>
    <kp>100000.0</kp>
    <kd>1.0</kd>
    <mu1>500.0</mu1>
    <mu2>200.0</mu2>
    <minDepth>0.002</minDepth>
    <maxVel>0</maxVel>
  </gazebo>
  <gazebo reference="left_gripper_link">
    <material>Gazebo/White</material>
    <selfCollide>true</selfCollide>
    <kp>100000.0</kp>
    <kd>1.0</kd>
    <mu1>500.0</mu1>
    <mu2>200.0</mu2>
    <minDepth>0.002</minDepth>
    <maxVel>0</maxVel>
  </gazebo>
</robot>